docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,371 @@
+  :func:`.recurrent` decorator as explained in the
+  :doc:`tutorial about rnns</rnn>`.
+* :meth:`.Brick._allocate`: you should implement this method to allocate the
+  shared variables (often representing parameters) of the brick. In blocks,
+  by convention, the built-in bricks allocate their shared variables with nan
+  values and we recommand you to do the same.
+* :meth:`.Brick._initialize`: you should implement this method to initialize
+  the shared variables of your brick. This method is called after the
+  allocation.
+* :meth:`.Brick._push_allocation_config`: you should consider overwriting
+  this method if you want to allocate the children bricks in a specific way.
+* :meth:`.Brick._push_initialization_config`: you should consider
+  overwriting this method if you want to initialize the children bricks in a
+  specific way. You should inherit from :class:`.Initializable` to initialize
+  the potential children bricks recursively.
+* :meth:`.Brick.get_dim`: implementing this function is useful if you want
+  to provide a simple way to get the dimensions of the inputs and outputs of
+  the brick.
+
+If you want to inherit from a specific brick, check its docstring to
+identify the particular methods to overwrite and the attributes to define.
+
+apply method
+~~~~~~~~~~~~
+
+The :meth:`apply` method listed above is probably the most
+important method of your brick because it is the one that actually takes
+theano tensors as inputs, process them and return output tensors. You should
+decorate it with the :func:`.application` decorator, which names variables
+and register auxiliary variables of the operation you implement.
+It is used as follows:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application(inputs=['input1', 'input2'], outputs=['output'])
+    ...     def apply(self, input1, input2)
+    ...         ...
+    ...         return something
+
+In the case above, it will automatically label the theano tensor variable
+``input1`` to ``Foo_apply_input1``, idem for ``input2`` and the output of the
+method.
+
+Under the hood, the ``@application`` decorator creates an object of class
+:class:`.Application`, named ``apply``, which becomes an attribute of the
+brick.
+
+Application properties
+""""""""""""""""""""""
+
+In the previous examples, variables were named with strings. If you need to
+name certain variables with other variables (such as ``self.fancy_name``),
+you should define them with the ``apply.property`` decorator:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     fancy_name = "salut_toi"
+    ...     @application
+    ...     def apply(self, input)
+    ...         ...
+    ...     @apply.property('inputs')
+    ...     def apply_inputs(self):
+    ...         # Note that you can use any python code to define the name
+    ...         return self.fancy_name
+
+Using application calls
+"""""""""""""""""""""""
+
+You can also annotate specific variables by passing ``application_call`` as
```

Comment:
The purpose of providing `application_call` is not "to annotate specific variables", but to allow adding auxiliary variables to the computation graph.